nav2_world_model/src/world_model.cpp

```diff
@@ -12,67 +12,110 @@
+  costmap_thread_ = std::make_unique<std::thread>(
+    [](rclcpp_lifecycle::LifecycleNode::SharedPtr node)
+    {rclcpp::spin(node->get_node_base_interface());}, costmap_ros_);
+}
+
+WorldModel::~WorldModel()
+{
+  RCLCPP_INFO(get_logger(), "Destroying");
+  costmap_thread_->join();
+}
+
+nav2_lifecycle::CallbackReturn
+WorldModel::onConfigure(const rclcpp_lifecycle::State & state)
 {
-  costmap_ros_ = std::make_shared<nav2_costmap_2d::Costmap2DROS>("global_costmap", tfBuffer_);
-  costmap_ = costmap_ros_->getCostmap();
-  executor.add_node(costmap_ros_);
+  RCLCPP_INFO(get_logger(), "onConfigure");
+
+  costmap_ros_->onConfigure(state);
 
   // Create a service that will use the callback function to handle requests.
-  costmapServer_ = create_service<nav2_msgs::srv::GetCostmap>("GetCostmap",
-      std::bind(&WorldModel::costmap_callback, this,
+  costmap_service_ = create_service<nav2_msgs::srv::GetCostmap>("GetCostmap",
+      std::bind(&WorldModel::costmap_service_callback, this,
       std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+WorldModel::onActivate(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_INFO(get_logger(), "onActivate");
+
+  costmap_ros_->onActivate(state);
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
 }
 
-void WorldModel::costmap_callback(
+nav2_lifecycle::CallbackReturn
+WorldModel::onDeactivate(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_INFO(get_logger(), "onDeactivate");
+
+  costmap_ros_->onDeactivate(state);
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+WorldModel::onCleanup(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_INFO(get_logger(), "onCleanup");
+
+  costmap_ros_->onCleanup(state);
+  costmap_service_.reset();
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+void
+WorldModel::costmap_service_callback(
   const std::shared_ptr<rmw_request_id_t>/*request_header*/,
   const std::shared_ptr<nav2_msgs::srv::GetCostmap::Request>/*request*/,
   const std::shared_ptr<nav2_msgs::srv::GetCostmap::Response> response)
 {
-  RCLCPP_INFO(this->get_logger(), "Received costmap request");
+  RCLCPP_INFO(get_logger(), "Received costmap service request");
+
+  // TODO(bpwilcox): Grab correct orientation information
+  tf2::Quaternion quaternion;
+  quaternion.setRPY(0.0, 0.0, 0.0);
+
+  nav2_costmap_2d::Costmap2D * costmap_ = costmap_ros_->getCostmap();
+
+  auto size_x = costmap_->getSizeInCellsX();
+  auto size_y = costmap_->getSizeInCellsY();
+  auto data_length = size_x * size_y;
+  unsigned char * data = costmap_->getCharMap();
 
-  response->map.metadata.size_x = costmap_->getSizeInCellsX();
-  response->map.metadata.size_y = costmap_->getSizeInCellsY();
+  response->map.header.stamp = now();
+  response->map.header.frame_id = "map";
+  response->map.metadata.size_x = size_x;
+  response->map.metadata.size_y = size_y;
```

Comment:
Do these changes have to do with Lifecycle nodes, or is this just some unrelated cleanup?